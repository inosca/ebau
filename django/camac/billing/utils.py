from collections import OrderedDict
from decimal import Decimal
from typing import List, TypedDict, Union

from camac.billing.models import BillingV2Entry


class OrganizationTotals(TypedDict):
    uncharged: str
    total: str


class BillingTotals(TypedDict):
    BillingV2Entry.CANTONAL: OrganizationTotals
    BillingV2Entry.MUNICIPAL: OrganizationTotals
    all: OrganizationTotals


def round_decimal(num: Decimal) -> Decimal:
    """Round decimal to 2 decimal places."""

    return num.quantize(Decimal("0.01"))


def calculate_final_rate(
    calculation: str,
    total_cost: Union[Decimal, None] = None,
    percentage: Union[Decimal, None] = None,
    hours: Union[Decimal, None] = None,
    hourly_rate: Union[Decimal, None] = None,
) -> Union[Decimal, None]:
    """Calculate final rate for given calculation type.

    - flat: use `total_cost`
    - percentage: `percentage` of `total_cost`
    - hourly: `hours` times `hourly_rate`
    """

    final_rate = None

    if calculation == BillingV2Entry.CALCULATION_FLAT:
        final_rate = total_cost
    elif calculation == BillingV2Entry.CALCULATION_PERCENTAGE:
        final_rate = total_cost * percentage / Decimal(100)
    elif calculation == BillingV2Entry.CALCULATION_HOURLY:
        final_rate = hours * hourly_rate

    # Don't ignore final_rate when value is 0
    return round_decimal(final_rate) if final_rate is not None else None


def add_taxes_to_final_rate(
    final_rate: Decimal, tax_mode: str, tax_rate: Decimal
) -> Union[Decimal, None]:
    """Add taxes to final rate.

    This only applies the `tax_rate` to the `final_rate` if the `tax_mode` is
    "exclusive".
    """

    # Don't ignore final_rate when value is 0
    if final_rate is None:
        return None

    if tax_mode != BillingV2Entry.TAX_MODE_EXCLUSIVE:
        return final_rate

    return round_decimal(final_rate + final_rate * tax_rate / Decimal(100))


def get_totals(entries: List[OrderedDict]) -> BillingTotals:
    """Get totals for a list of billing entries.

    This will return a dict of totals per organization type and over all
    organizations (including entries without an organization).
    """

    totals = {}

    for key, _ in BillingV2Entry.ORGANIZATION_CHOICES:
        totals[key] = get_totals_for_organization(entries, key)

    totals["all"] = get_totals_for_organization(entries)

    return totals


def get_totals_for_organization(
    entries: List[OrderedDict], organization: Union[str, None] = None
) -> OrganizationTotals:
    """Calculate total including uncharged total for an organization."""

    filtered_entries = [
        entry
        for entry in entries
        if not organization or entry["organization"] == organization
    ]

    return {
        "uncharged": str(
            round_decimal(
                Decimal(
                    sum(
                        [
                            Decimal(entry["final_rate"])
                            for entry in filtered_entries
                            if entry["date_charged"] is None
                            and entry["final_rate"]
                            is not None  # Don't ignore entry when final_rate is 0
                        ]
                    )
                )
            )
        ),
        "total": str(
            round_decimal(
                Decimal(
                    sum(
                        [
                            Decimal(entry["final_rate"])
                            for entry in filtered_entries
                            if entry["final_rate"]
                            is not None  # Don't ignore entry when final_rate is 0
                        ]
                    )
                )
            )
        ),
    }
