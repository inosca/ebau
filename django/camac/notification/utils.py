from collections import namedtuple

from camac.notification.models import NotificationTemplate
from camac.notification.serializers import (
    NotificationTemplateSendmailSerializer,
    PermissionlessNotificationTemplateSendmailSerializer,
)
from camac.user.models import Group, User

Request = namedtuple("Request", ["user", "group", "query_params"])


def send_mail_without_request(slug, username=None, group_id=None, **kwargs):
    """Send notification email if you don't have a HTTP request.

    Note: You can leave out username and group_id. In that case, the emails
    will be sent from the system / support account.
    """
    if not username or not group_id:
        context = {}
    else:
        context = {
            "request": Request(
                user=User.objects.get(username=username),
                group=Group.objects.get(pk=group_id),
                query_params=[],
            )
        }

    return send_mail(
        slug,
        context,
        serializer=PermissionlessNotificationTemplateSendmailSerializer,
        **kwargs,
    )


def send_mail(
    slug,
    context,
    serializer=NotificationTemplateSendmailSerializer,
    **kwargs,
):
    """Call a SendmailSerializer based on a NotificationTemplate Slug."""
    notification_template = NotificationTemplate.objects.get(slug=slug)

    data = {
        "notification_template": {
            "type": "notification-templates",
            "id": notification_template.pk,
        },
        **kwargs,
    }

    serializer = serializer(data=data, context=context)
    serializer.is_valid(raise_exception=True)
    serializer.save()

    return serializer
