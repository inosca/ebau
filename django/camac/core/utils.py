from typing import Callable

from caluma.caluma_workflow.models import Case
from django.conf import settings
from django.db.models import CharField, IntegerField
from django.db.models.fields.json import KeyTextTransform
from django.db.models.functions import Cast, Substr
from django.utils import timezone

from camac.core.models import HistoryActionConfig
from camac.core.translations import get_translations
from camac.instance.models import HistoryEntry, HistoryEntryT, Instance
from camac.user.models import User


def generate_ebau_nr(year: int) -> str:
    max_number = (
        Case.objects.filter(**{"meta__ebau-number__startswith": year})
        .annotate(
            ebau_nr=Cast(
                Substr(Cast(KeyTextTransform("ebau-number", "meta"), CharField()), 6),
                IntegerField(),
            )
        )
        .order_by("-ebau_nr")
        .values_list("ebau_nr", flat=True)
        .first()
    )

    return "%d-%d" % (year, (max_number or 0) + 1)


def assign_ebau_nr(instance: Instance, year=None) -> str:
    existing = instance.case.meta.get("ebau-number")
    if existing:
        return existing

    year = timezone.now().year if year is None else year

    ebau_nr = generate_ebau_nr(year)

    instance.case.meta["ebau-number"] = ebau_nr
    instance.case.save()

    return ebau_nr


def create_history_entry(
    instance: Instance,
    user: User,
    text: str,
    text_data: Callable[[str], dict] = lambda language: {},
    history_type: str = HistoryActionConfig.HISTORY_TYPE_STATUS,
    body: str = "",
) -> None:
    """
    Create a multilingual history entry for an instance.

    The parameters `instance`, `user` and `text` are required. Optionally it
    accepts a function `text_data` that takes the language as parameter and
    should return a dictionary of data that will be formatted into the text.
    Also, the `history_type` can be passed if it's anything other than a
    status change.

    >>> create_history_entry(
    ...     instance,
    ...     user,
    ...     gettext_noop("Your text"),
    ...     text_data_function,
    ...     history_type
    ... )
    """

    data = {}
    if not settings.APPLICATION.get("IS_MULTILINGUAL", False):
        data = {"title": text, "body": body}

    history = HistoryEntry.objects.create(
        instance=instance,
        created_at=timezone.now(),
        user=user,
        history_type=history_type,
        **data,
    )

    for (language, text) in get_translations(text).items():
        HistoryEntryT.objects.create(
            history_entry=history,
            title=text % text_data(language),
            body=body,
            language=language,
        )
