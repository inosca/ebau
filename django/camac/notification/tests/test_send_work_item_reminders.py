from datetime import timedelta

import pytest
from django.core.management import call_command
from django.utils import timezone


@pytest.mark.freeze_time("2020-08-10")
@pytest.mark.parametrize("multilingual", [True, False])
@pytest.mark.parametrize(
    "is_overdue,is_not_viewed,is_assigned,has_controlling,multi_mail_service,is_applicant,outbox_count",
    [
        (True, True, True, True, False, False, 3),
        (True, True, True, False, True, False, 3),
        (True, False, True, False, False, False, 2),
        (False, True, True, False, True, False, 3),
        (False, True, False, False, False, False, 1),
        (False, False, True, False, False, False, 0),
        (True, True, True, True, False, True, 2),
        (True, False, True, False, False, True, 1),
    ],
)
def test_send_work_item_reminders(
    application_settings,
    db,
    mailoutbox,
    be_instance,
    work_item_factory,
    task_factory,
    snapshot,
    service_factory,
    service_t_factory,
    user_factory,
    is_overdue,
    is_not_viewed,
    is_assigned,
    has_controlling,
    multi_mail_service,
    is_applicant,
    outbox_count,
    multilingual,
):
    application_settings["IS_MULTILINGUAL"] = multilingual

    user = user_factory()
    services = service_factory.create_batch(2)

    if multilingual:
        for service in services:
            for language in ["de", "fr"]:
                service_t_factory(language=language, service=service)

    if multi_mail_service:
        services[0].email = f"{services[0].email},foo@bar.com"
        services[0].save()

    deadline = (
        timezone.now() - timedelta(days=1)
        if is_overdue
        else timezone.now() + timedelta(days=1)
    )

    work_item_factory(
        status="ready",
        meta={"not-viewed": is_not_viewed},
        deadline=deadline,
        assigned_users=[user.username] if is_assigned else [],
        addressed_groups=[str(services[0].pk)] if not is_applicant else ["applicant"],
        controlling_groups=[str(services[1].pk)] if has_controlling else [],
    )

    call_command("send_work_item_reminders")

    assert len(mailoutbox) == outbox_count
    snapshot.assert_match(
        [(mail.subject, mail.body, mail.to, mail.cc) for mail in mailoutbox]
    )


@pytest.mark.parametrize("user__disabled", [1])
def test_dont_send_reminders_caluma(db, user, service, work_item_factory, mailoutbox):
    service.disabled = 1
    service.save()
    work_item_factory(
        status="ready",
        meta={"not-viewed": True},
        deadline=timezone.now() - timedelta(days=1),
        assigned_users=[user.username],
        addressed_groups=[service.pk],
        controlling_groups=[service.pk],
    )
    call_command("send_work_item_reminders")
    assert len(mailoutbox) == 0
