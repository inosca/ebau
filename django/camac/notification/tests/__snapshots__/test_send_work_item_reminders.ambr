# serializer version: 1
# name: test_send_work_item_reminders[False-False-True-False-False-False-0-False]
  list([
  ])
# ---
# name: test_send_work_item_reminders[False-False-True-False-False-False-0-True]
  list([
  ])
# ---
# name: test_send_work_item_reminders[False-True-False-False-False-False-1-False]
  list([
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Jeffrey Zhang) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[False-True-False-False-False-False-1-True]
  list([
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Sarah Santiago) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Derek Wagner) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 0 tâches de contrôle en retard
        - 1 tâche de contrôle non lue
        - 0 tâches en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[False-True-True-False-True-False-3-False]
  list([
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Jeffrey Zhang) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Jeffrey Zhang) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'foo@bar.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[False-True-True-False-True-False-3-True]
  list([
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour Rebecca Gonzalez,
        
        Vous avez les tâches suivantes dans eBau qui requièrent votre attention :
        
        - 0 tâches de contrôle en retard
        - 1 tâche de contrôle non lue
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Sarah Santiago) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Derek Wagner) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 0 tâches de contrôle en retard
        - 1 tâche de contrôle non lue
        - 0 tâches en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Sarah Santiago) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Derek Wagner) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 0 tâches de contrôle en retard
        - 1 tâche de contrôle non lue
        - 0 tâches en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'foo@bar.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-False-True-False-False-False-2-False]
  list([
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 0 ungelesene Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Jeffrey Zhang) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 0 ungelesene Aufgaben
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-False-True-False-False-False-2-True]
  list([
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 0 ungelesene Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour Rebecca Gonzalez,
        
        Vous avez les tâches suivantes dans eBau qui requièrent votre attention :
        
        - 1 tâche de contrôle en retard
        - 0 tâches de contrôle non lues
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Sarah Santiago) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 0 ungelesene Aufgaben
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Derek Wagner) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 1 tâche de contrôle en retard
        - 0 tâches de contrôle non lues
        - 0 tâches en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-False-True-False-False-True-1-False]
  list([
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 0 ungelesene Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-False-True-False-False-True-1-True]
  list([
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 0 ungelesene Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour Rebecca Gonzalez,
        
        Vous avez les tâches suivantes dans eBau qui requièrent votre attention :
        
        - 1 tâche de contrôle en retard
        - 0 tâches de contrôle non lues
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-True-True-False-True-False-3-False]
  list([
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Jeffrey Zhang) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Jeffrey Zhang) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'foo@bar.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-True-True-False-True-False-3-True]
  list([
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour Rebecca Gonzalez,
        
        Vous avez les tâches suivantes dans eBau qui requièrent votre attention :
        
        - 1 tâche de contrôle en retard
        - 1 tâche de contrôle non lue
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Sarah Santiago) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Derek Wagner) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 1 tâche de contrôle en retard
        - 1 tâche de contrôle non lue
        - 0 tâches en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Sarah Santiago) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Derek Wagner) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 1 tâche de contrôle en retard
        - 1 tâche de contrôle non lue
        - 0 tâches en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'foo@bar.com',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-True-True-True-False-False-3-False]
  list([
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Jeffrey Zhang) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Maria Fuentes) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 0 ungelesene Aufgaben
        - 1 überfällige Controlling-Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'calebsoto@example.net',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-True-True-True-False-False-3-True]
  list([
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour Rebecca Gonzalez,
        
        Vous avez les tâches suivantes dans eBau qui requièrent votre attention :
        
        - 1 tâche de contrôle en retard
        - 1 tâche de contrôle non lue
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Sarah Santiago) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        - 0 überfällige Controlling-Aufgaben
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Derek Wagner) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 1 tâche de contrôle en retard
        - 1 tâche de contrôle non lue
        - 0 tâches en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'nicolefields@example.com',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Ana Schneider) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 0 ungelesene Aufgaben
        - 1 überfällige Controlling-Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Brandi Vargas) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 0 tâches de contrôle en retard
        - 0 tâches de contrôle non lues
        - 1 tâche en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'calebsoto@example.net',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-True-True-True-False-True-2-False]
  list([
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben',
      '''
        Guten Tag
        
        Ihre Organisation (Maria Fuentes) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 0 ungelesene Aufgaben
        - 1 überfällige Controlling-Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
  
      ''',
      list([
        'calebsoto@example.net',
      ]),
      list([
      ]),
    ),
  ])
# ---
# name: test_send_work_item_reminders[True-True-True-True-False-True-2-True]
  list([
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag Rebecca Gonzalez
        
        Sie haben folgende Aufgaben in eBau, welche Ihre Aufmerksamkeit benötigen:
        
        - 1 überfällige Aufgabe
        - 1 ungelesene Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour Rebecca Gonzalez,
        
        Vous avez les tâches suivantes dans eBau qui requièrent votre attention :
        
        - 1 tâche de contrôle en retard
        - 1 tâche de contrôle non lue
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'fmalone@example.org',
      ]),
      list([
      ]),
    ),
    tuple(
      'Erinnerung an Aufgaben / Rappel des tâches',
      '''
        Guten Tag
        
        Ihre Organisation (Ana Schneider) hat folgende Aufgaben in eBau, welche Aufmerksamkeit benötigen:
        
        - 0 überfällige Aufgaben
        - 0 ungelesene Aufgaben
        - 1 überfällige Controlling-Aufgabe
        
        http://ebau.local
        
        Diese E-Mail wurde automatisch generiert, bitte antworten Sie nicht darauf.
        
        *** version française ***
        
        Bonjour,
        
        Votre organisation (Brandi Vargas) a les tâches suivantes dans eBau qui requièrent une attention particulière :
        
        - 0 tâches de contrôle en retard
        - 0 tâches de contrôle non lues
        - 1 tâche en retard
        
        http://ebau.local
        
        Ce message a été généré automatiquement, veuillez ne pas y répondre.
  
      ''',
      list([
        'calebsoto@example.net',
      ]),
      list([
      ]),
    ),
  ])
# ---
